### PR TITLE
Rollup enable by default

### LINKED ISSUE #102: Have `rollup` enabled by default
## Context

Having rollup disabled is the exception. We should reflect that in the command line arguments to `cartesi-machine`.

## Possible solutions

Before: rollup **disabled** by default, `--rollup` enables it.
After: rollup **enabled** by default, `--no-rollup` disables it.

Something to consider is the deprecation process.
Tools or scripts that have `--rollup` would like a time period to update without completely being broken after an update.
Something to consider.

### PR DESCRIPTION
**Pull request by vfusco**
closes #102

### CODE REVIEW

File: src/cartesi-machine.lua [hunk 1]
```diff
@@ -607,33 +611,28 @@ local options = {
         end,
     },
     {
-        "^%-%-htif%-yield%-manual$",
+        "^%-%-no%-htif%-yield%-manual$",
         function(all)
             if not all then return false end
-            htif_yield_manual = true
+            htif_yield_manual = false
```
[the remark this answers is not in the document]
Comment (vfusco) [reply src/cartesi-machine.lua:617]: Sure it won't work. The user will get an error while trying to use it. But isn't this true for all other misconfigured parameters? So I'm not sure I agree, as yield can be used and was used in the past without rollups.

I was in doubt if I should disable the `htif_yield_manual` and `htif_yield_automatic` on `--no-rollup`. Now I think I shouldn't as there is no other way to turn it on if the users want to.

File: src/cartesi-machine.lua [hunk 2]
```diff
@@ -607,33 +611,28 @@ local options = {
...
             return true
         end,
     },
     {
-        "^%-%-htif%-yield%-automatic$",
+        "^%-%-no%-htif%-yield%-automatic$",
         function(all)
             if not all then return false end
-            htif_yield_automatic = true
+            htif_yield_automatic = false
```
[the remark this answers is not in the document]
Comment (vfusco) [reply src/cartesi-machine.lua:625]: see previous comment

File: src/cartesi-machine.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpolitzer) [reply src/cartesi-machine.lua:617]: > I was in doubt if I should disable the htif_yield_manual and htif_yield_automatic on --no-rollup. Now I think I shouldn't as there is no other way to turn it on if the users want to.

This is what will happen after the PR got updated:
`--no-rollup` -> rollup driver won't load due
`--no-yield-automatic` -> yield with `progress` and rollup device will be broken (can `finish` but can't `write`).
`--no-yield-manual` -> rollup device will be broken (can `write` but can't `finish`).

We could require both yields types to be available in the rollup driver initialization.
The yield flags are available in the FDT and so something limilar to what is in the yield driver.

There is also this idea to limit the yield driver to only allow `progress` calls.
Other calls can only be performed via the rollup driver.
Currently, with yield access, a process can mess up the rollup protocol by doing random yields.

https://github.com/cartesi/linux/pull/4/commits/283657c4ea8fa85dd48ee6610961ac4e714a3396

[the remark this answers is not in the document]
Comment (diegonehab) [reply src/cartesi-machine.lua:617]: If the user disables either of the yields in the command-line, but does not disable rollups as well, we will end up with a machine config that has rollup but does not have yields. The emulator should throw and exception from within the constructor of `machine` explaining the problem. Right?

[the remark this answers is not in the document]
Comment (diegonehab) [reply src/cartesi-machine.lua:617]: There is a weirdness where the user enables the yield before disabling rollup... That will be hard to catch. Can we fix the issue as well?

File: src/cartesi-machine.lua [hunk 3]
```diff
@@ -448,7 +446,13 @@ local init_splash = true
 local append_bootargs = ""
 local append_init = ""
 local append_entrypoint = ""
-local rollup
+local rollup = {
```
Comment (edubart) [inline src/cartesi-machine.lua:449]: This only solves for `cartesi-machine.lua`, what about the C API `cm_get_default_config` ? Shoudn't we also make this the default there? cc @diegonehab

File: src/machine-config.h [hunk 4]
```diff
@@ -112,17 +112,17 @@ struct htif_config final {
...
 
 /// \brief Rollup configuration
-struct rollup_config {                    // NOLINT(bugprone-exception-escape)
-    memory_range_config rx_buffer{};      ///< RX buffer
-    memory_range_config tx_buffer{};      ///< TX buffer
-    memory_range_config input_metadata{}; ///< Buffer for input metadata
-    memory_range_config voucher_hashes{}; ///< Buffer for the voucher hash array
-    memory_range_config notice_hashes{};  ///< Buffer for the notice hash array
+struct rollup_config {                                                          // NOLINT(bugprone-exception-escape)
+    memory_range_config rx_buffer{.start = 0x60000000, .length = 2 << 20};      ///< RX buffer
```
Comment (edubart) [inline src/machine-config.h:121]: It compiles, but I am getting these warnings:
```c++
machine-config.h:121:35: warning: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Wc++20-extensions]
  121 |     memory_range_config rx_buffer{.start = 0x60000000, .length = 2 << 20};      ///< RX buffer
      |                                   ^
machine-config.h:121:56: warning: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Wc++20-extensions]
  121 |     memory_range_config rx_buffer{.start = 0x60000000, .length = 2 << 20};      ///< RX buffer
      |                                                        ^
```

Using `memory_range_config rx_buffer{0x60000000, 2 << 20};` should fix.

**Review by edubart — CHANGES_REQUESTED** [review]

File: src/machine-config.h (hunk 4 above)
Comment (diegonehab) [reply src/machine-config.h:121]: I used it in LambaDEX, but I selected c++20. Victor must have seen it. Shall we make the switch? This is really cool. :)

File: src/cartesi-machine.lua (hunk 3 above)
Comment (diegonehab) [reply src/cartesi-machine.lua:449]: Yes. It should be default in C++ as well.

Comment (diegonehab) [reply src/cartesi-machine.lua:449]: Victor will create an issue about 'get_default_config' vs. 'cartesi-machine.lua'

**Review by diegonehab — APPROVED** [review]

File: src/cartesi-machine.lua (hunk 3 above)
Comment (vfusco) [reply src/cartesi-machine.lua:449]: https://github.com/cartesi/machine-emulator/issues/153

**Pull request merged**